README.md
# mctext

A simple text generator based on Markov Chains I built a long time ago.

## Building

```
./autogen.sh
./configure && make
```

## Usage

```
Usage: mctext [options] FILE
Reads a sample text from file and generate new text based on it using Markov chains.

Options:
  -h [ --help ]                   produce this help message
  --version                       prints version string
  -w [ --num-of-words ] arg (=40) how many words should be generated
  -s [ --steps ] arg (=2)         the number of steps used in the Markov chain.

With no FILE, or when FILE is -, read standard input.
```







ChangeLog
mctext-0.1:
-----------
- Initial version.


text_generator.h
/***************************************************************************
 *   Copyright (C) 2008 by Guy Rutenberg                                   *
 *   guyrutenberg@gmail.com                                                *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#ifndef TEXT_GENERATOR_H
#define TEXT_GENERATOR_H

#include <fstream>
#include <string>
#include <vector>

class TextGenerator {
public:
	TextGenerator();
	void addWords(const std::string &str);
	std::string generateWords(int count, int steps);
private:
	std::string stripChars(std::string str);
	std::vector<std::string> m_words;

};

#endif //TEXT_GENERATOR_H


text_generator.cpp
/***************************************************************************
 *   Copyright (C) 2008 by Guy Rutenberg                                   *
 *   guyrutenberg@gmail.com                                                *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#include <fstream>
#include <string>
#include <vector>
#include <sstream>
#include <algorithm>
#include <iostream>
#include <boost/random/uniform_int.hpp>
#include <boost/random/linear_congruential.hpp>
#include <boost/random/variate_generator.hpp>

#include "text_generator.h"

using namespace std;

/**
 *\param sample_file an open stream to a file holding sample text.
 */
TextGenerator::TextGenerator()
{
	//TODO initialize the random number generator here;
}

/**
 * This function takes a string of words, and adds them into the sample text
 * base
 */
void TextGenerator::addWords(const std::string &str)
{
	string buf;
	stringstream ss(str);

	while (ss >> buf)
		m_words.push_back(buf);
	
}


string TextGenerator::generateWords(int count, const int steps)
{
	const int vec_length = m_words.size();

	string output;
	vector<string> temp;
	
	//TODO move the random number generator from here to a class member
	boost::uniform_int<size_t> uni_dist(0, vec_length-steps);
	boost::variate_generator<boost::rand48, boost::uniform_int<size_t> >
		rand(boost::rand48((int) time(NULL)), uni_dist);

	//this is for randomizing the search order
	boost::uniform_int<size_t> uni_int(0, -1);
	boost::variate_generator<boost::rand48, boost::uniform_int<size_t> >
		randint(boost::rand48((int) time(NULL)), uni_int);
	//TODO check if we can give up on one of the random number generators
	
	int initial_rand = rand();
	for (int i=0; i<steps && count>0; ++i) {
		temp.push_back(m_words[initial_rand+i]);	
		--count;
	}
	vector< vector<string>::iterator > itr_vec; //this will hold iterator to results

	for (int i = count; i>0 ; --i) {
		itr_vec.clear();
		vector<string>::iterator last_it = m_words.begin();

		last_it = search(last_it, m_words.end()-1, temp.end()-steps, temp.end());

		while (last_it != m_words.end()) {
			itr_vec.push_back(last_it+steps);
			last_it = search(last_it+1, m_words.end(), temp.end()-steps, temp.end());
		}
		if (itr_vec.size()==0) { 
			cerr<<"debug: don't know how to continue - "<<temp.size()<<endl;
			initial_rand = rand();
			for (int j=0; i<steps, i>0; ++j) {
				temp.push_back(m_words[initial_rand+i]);	
				--i;
			}
			continue;
		}
		temp.push_back( *(itr_vec[ randint()%itr_vec.size() ] ) ); //there is probably better way to code this
	}
	
	for (int i = 0; i < temp.size(); ++i) {
		output += temp[i]+" ";
	}
	return output;
}


main.cpp
/***************************************************************************
 *   Copyright (C) 2008 by Guy Rutenberg                                   *
 *   guyrutenberg@gmail.com                                                *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#include <iostream>
#include <fstream>
#include <boost/program_options.hpp>
#include "text_generator.h"
#include "config.h"

namespace po = boost::program_options;
using namespace std;

string usage();

int main (int argc, char **argv)
{
	po::options_description desc("Options");
        desc.add_options()
		("help,h", "produce this help message")
		("version", "prints version string")
		("num-of-words,w", po::value<int>()->default_value(40), "how many words should be generated")
		("steps,s", po::value<int>()->default_value(2), "the number of steps used in the Markov chain.")
        ;

	po::options_description hidden("Hidden options");
	hidden.add_options()
		("input-file", po::value<string>()->default_value("-"), "sample text input file")
	;

	po::options_description cmdline_options;
	cmdline_options.add(desc).add(hidden);

	po::positional_options_description p;
	p.add("input-file", 1);

	po::variables_map vm;
	try {
		po::store(po::command_line_parser(argc, argv).
			options(cmdline_options).positional(p).run(), vm);
	}
	catch(po::too_many_positional_options_error &e)
	{
		cerr<<e.what()<<endl;
		cerr<<usage();
		return 1;
	}
	catch(po::unknown_option &e)
	{
		cerr<<PACKAGE<<": "<<e.what()<<endl;
		cerr<<usage();
		cerr<<"Try `"<<PACKAGE<<" --help' for more information."<<endl;
		return 1;
	}
	po::notify(vm);
	if (vm.count("help")) {
		cout<<usage();
		cout<<"Reads a sample text from file and generate new text based on it using Markov chains.\n\n";
		cout << desc << "\n";
		cout <<"With no FILE, or when FILE is -, read standard input.\n"<<endl;
		cout <<"Report bugs to <"<< PACKAGE_BUGREPORT <<">"<<endl;

		return 0;
	}
	if (vm.count("version")) {
		cout << PACKAGE_STRING << "\n";
		return 0;
	}

	TextGenerator gen;
	// read file (or stdin) and generate text
	if (vm["input-file"].as<string>() == "-") {
		string line;
		while (!cin.eof()) {
			getline(cin, line);
			gen.addWords(line);
		}
	} else {
		ifstream file (vm["input-file"].as<string>().c_str());
		if (!file.is_open()) {
			cerr<<"Error openning file: "<<vm["input-file"].as<string>()<<endl;
			return 1;
		}
		string line;
		while (!file.eof()) {
			getline(file, line);
			gen.addWords(line);
		}
		file.close();
	}

	int num_words = vm["num-of-words"].as<int>();
	int num_steps = vm["steps"].as<int>();
	cout<<gen.generateWords(num_words, num_steps)<<endl;

	return 0;
}

/**
 * Prints usage information.
 */
string usage()
{
	string out = "Usage: ";
	out += PACKAGE;
	out += " [options] FILE\n";
	return out;
}
